
---
title: "Comparing OC8901 and OC8345 - wt only (Hernan Revision)"
author: "Amanda Yeo, Hernan Lorenzi, Orna Cohen-Fix"
date: "2023-11-16"
output: html_document
---

````{r include=FALSE}
knitr::opts_chunk$set(message = FALSE, 
                      warning = FALSE,
                      tidy.opts=list(width.cutoff=80),
                      tidy=TRUE
                      )
```

### Install required packages

```{r}
### Install required R libraries/packages:

# CRAN repository:
required_cran_packages <- c("cowplot","ggpubr","ggsci","RColorBrewer","pheatmap")
new_cran_packages <- required_cran_packages[!(required_cran_packages %in%
                                                installed.packages()[,"Package"])]
if(length(new_cran_packages)>0) {install.packages(new_cran_packages)}

# Bioconductor repository:
required_bioconductor_packages <- c("DESeq2","org.Sc.sgd.db","EnhancedVolcano","enrichplot","clusterProfiler")

new_bioconductor_packages <- required_bioconductor_packages[!(required_bioconductor_packages %in%
                                                installed.packages()[,"Package"])]

if(length(new_bioconductor_packages)>0) {
  if (!require("BiocManager", quietly = TRUE)){
    install.packages("BiocManager")
  }
  BiocManager::install(new_bioconductor_packages)
}
```

### Load R libraries

```{r}
# Packages for differential expression analysis and plotting:

suppressWarnings(
  suppressPackageStartupMessages({
    library(DESeq2);  # Package for differential expression analysis
    library(cowplot); # Package for saving plots (ggsave2)
    library(ggpubr); # Package for plotting functions
    library(ggsci);
    library(RColorBrewer); 
    library(pheatmap); # Package for volcane plot functions
    library(org.Sc.sgd.db); # Yeast annotations
    library(EnhancedVolcano);
    library(clusterProfiler);
    library(enrichplot);
    }
  )
)
# Packages for functional analysis:

#library(clusterProfiler)
#library(enrichplot)
#library(ggupset)
#library(msigdbr)
#library(ReactomePA)
#library(ggraph)
#library(BSgenome)
```

### Load useful functions

```{r}
#
# --- function for adding gene names to DESeq2 result table ---
#
get_gene_names_from_gene_ids <- function(orf_ids, annotation_db, 
                                         look_for = 'ORF', 
                                         fetch = 'GENENAME', return_all = TRUE){
  # Reference organism: Saccharomyces cerevisiae => DATABASE = org.Sc.sgd.db
  symbols <- mapIds(annotation_db, keys = orf_ids, column = fetch, 
                    keytype = look_for, multiVals = "first")
  symbols <- symbols[!is.na(symbols)]
  to_name <- orf_ids %in% names(symbols)
  orf_ids[to_name] <- as.vector(symbols)
  if (return_all){
    return(orf_ids)
  }
  else {
    return(orf_ids[to_name])
  }
}

#
# --- function for PCA plots ---
#
plot_PCA = function(object, color_by="condition", 
                    shape_by = 19, ntop=500, size = 3,
                    returnData=FALSE, pcs = c(1,2))
#lty <- c("solid", "dashed", "dotted", "dotdash", "longdash", "twodash")
#linetypes <- data.frame(
 # y = seq_along(lty),
  #lty = lty
#) 
{
  # Check variables are present in object
  intgroup = c(color_by)
  if (shape_by != 19){intgroup <- c(intgroup, shape_by)}
  if (!all(intgroup %in% names(colData(object)))) {
    stop("the argument 'intgroup' should specify columns of colData(dds)")
  }

  # calculate the variance for each gene
  rv <- rowVars(assay(object))

  # select the ntop genes by variance
  select <- order(rv, decreasing=TRUE)[seq_len(min(ntop, length(rv)))]

  # perform a PCA on the data in assay(x) for the selected genes
  pca <- prcomp(t(assay(object)[select,]))

  # the contribution to the total variance for each component
  percentVar <- pca$sdev^2 / sum( pca$sdev^2 )

  
  intgroup.df <- as.data.frame(colData(object)[, intgroup, drop=FALSE])
  
  # add the intgroup factors together to create a new grouping factor
  group <- if (length(intgroup) > 1) {
    factor(apply( intgroup.df, 1, paste, collapse=":"))
  } else {
    colData(object)[[intgroup]]
  }

  # assembly the data for the plot
  d <- data.frame(PC1=pca$x[,pcs[1]], PC2=pca$x[,pcs[2]], group=group, intgroup.df, name=colnames(object))
  colnames(d)[1] <- paste0("PC",pcs[1])
  colnames(d)[2] <- paste0("PC",pcs[2])
  
  if (returnData) {
    attr(d, "percentVar") <- percentVar[1:2]
    return(d)
  }
  
  ggplot(data=d, aes_string(x=colnames(d)[1], y=colnames(d)[2], color=color_by, shape=shape_by)) + 
    geom_point(size=size) + 
    scale_color_lancet() +
    xlab(paste0("PC",pcs[1],": ",round(percentVar[pcs[1]] * 100),"% variance")) + # fixed
    ylab(paste0("PC",pcs[2],": ",round(percentVar[pcs[2]] * 100),"% variance")) + # fixed
    coord_fixed(ratio = (max(d[,1])-min(d[,1]))/(max(d[,2])-min(d[,2])))
}

#
# --- function to plot gene expression of top10 genes ---
#
# Plot top10 diff expressed genes
plot_expression_topN_genes <- function(deseq_result, N=10){
  my_top10 <- head(rownames(deseq_result),N)
  for (gene in my_top10){
    print(gene)
    plotCounts(dds, gene, 
               intgroup = c("treatment","genotype"), 
               returnData = FALSE, col=metadata$genotype)
  }
}
```

### Load data

Hernan's code below: Basically the same as Amanda's but loads read counts and metadata for all samples and then split the tables into  genotype-specific input files.
```{r}
# Import read counts table into read_counts variable
read_counts.all <- read.table(file = "read_counts_table_comp8901vs8345.txt",  
                          header = TRUE, # First line contains column headers 
                          row.names = 1, # Column 1 contains row names
                          sep = "\t")   # Column delimiter is a tab (\t)

# Round read counts to the closest integer
read_counts.all <- round(read_counts.all, digits = 0)

# Read metadata.all table
metadata.all <- read.table(file = "metadata_OC8901vsOC8345.txt", 
                       header = TRUE, 
                       row.names = 1,
                       sep = "\t")

# Sort read_counts.all table so metadata.all and read_counts.all match order
if(!unique(colnames(read_counts.all) == rownames(metadata.all))){
  read_counts.all <- read_counts.all[, match(rownames(metadata.all), colnames(read_counts.all))]
}

# Include sample ids in metadata.all as a variable (column)
metadata.all$sample_id <- c("AY31", "AY32" , "AY33" , "AY34" , "AY35" , "AY36" , "AY37" , "AY38" , "AY39" , "AY40" , "AY41" , "AY42", "AY1", "AY2", "AY3", "AY4","AY5", "AY6", "AY7", "AY8", "AY9", "AY10" , "AY11" , "AY12")

# Include total read counts in metadata.all
metadata.all$read_counts <- colSums(read_counts.all, na.rm = TRUE)

# Make group, strain, treatment and genotype columns as factors
metadata.all$treatment <- as.factor(metadata.all$treatment)
metadata.all$strain <- as.factor(metadata.all$strain)
metadata.all$genotype <- as.factor(metadata.all$genotype)
metadata.all$experiment <- as.factor(metadata.all$experiment)

# Keep only WT-specific data to match Amanda's code inputs
keep <- metadata.all$genotype=="wt"
metadata <- metadata.all[keep,]
read_counts <- read_counts.all[,keep]
```

Amanda's code below
```{r}
# # Import read counts table into read_counts variable
# read_counts <- read.table(file = "read_counts_table_comp8901vs8345_WTonly.txt",  
#                           header = TRUE, # First line contains column headers 
#                           row.names = 1, # Column 1 contains row names
#                           sep = "\t")   # Column delimiter is a tab (\t)
# 
# # Round read counts to the closest integer
# read_counts <- round(read_counts, digits = 0)
# 
# # Read metadata table
# metadata <- read.table(file = "metadata_OC8901vsOC8345_WTonly.txt", 
#                        header = TRUE, 
#                        row.names = 1,
#                        sep = "\t")
# 
# # Sort read_counts table so metadata and read_counts match order
# if(!unique(colnames(read_counts) == rownames(metadata))){
#   read_counts <- read_counts[, match(rownames(metadata), colnames(read_counts))]
# }
# 
# # Include sample ids in metadata as a variable (column)
# metadata$sample_id <- c("AY39", "AY40" , "AY41" , "AY42" , "AY9", "AY10", "AY11", "AY12")
# 
# # Include total read counts in metadata
# metadata$read_counts <- colSums(read_counts, na.rm = TRUE)
# 
# # Make group, strain, treatment and genotype columns as factors
# metadata$treatment <- as.factor(metadata$treatment)
# metadata$strain <- as.factor(metadata$strain)
# metadata$genotype <- as.factor(metadata$genotype)
# metadata$experiment <- as.factor(metadata$experiment)

```

### DE analysis with DESeq2

```{r}
# Create folder for storing plots
dir.create(path = "./Plots_OC8901vsOC8345_wt", showWarnings = FALSE)

# Add extra metadata column with treatment and genotype data combined
metadata$treat_geno <-  paste(metadata$treatment, metadata$genotype, sep = '_')

# Create DESeq2 object
dds <- DESeqDataSetFromMatrix(countData = read_counts,
                              colData = metadata,
                              design = ~ experiment + treat_geno) # Here we assume that all genotypes respond the same to treatment


# Make sure reference groups are "none" and "wt". This means those categories will be the denominator when calculating Log2FC.
dds$treatment <- relevel(dds$treatment, "none")
#dds$genotype <- relevel(dds$genotype, "wt")
dds$treat_geno <- relevel(dds$treat_geno, "none_wt") 
dds <- DESeq(dds)
```
### Explortory analysis #1

```{r}
# Plot total reads per sample using barchar
readcounts.p <- ggbarplot(data = metadata,
                x = "sample_id",
                y = "read_counts",
                x.text.angle = 90,
                fill = "treatment",
                title = "Total read counts per sample",
                ylab = "Read counts",
                sort.by.groups = TRUE,
                palette = c("red","orange"), #"jco",
                sort.val = "asc") + xlab("Sample") + ylab("Read counts")
ggsave2("Plots_OC8901vsOC8345_wt/barplot_read_counts_per_sample_AMANDA.pdf", plot = readcounts.p)

print(readcounts.p)
```
**Figure 1:** Number of reads per sample

### Preprocess data stored in DESeq2 object dds

```{r}
# 1. Normalize counts
dds.vst <- vst(dds, blind=TRUE)

# 2. Keep genes with at least 10 reads total across all samples
keep <- rowSums(as.data.frame(dds.vst@assays@data@listData)) >= 10
dds.vst <- dds.vst[keep,]

# 3. Calculate distances between samples
sampleDists <- dist(t(assay(dds.vst)))

# 4. Plot inter-sample distances
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(rownames(sampleDistMatrix), dds.vst$treatment)

#colnames(sampleDistMatrix) <- NULL
hm.p <- pheatmap(mat = sampleDistMatrix,
                clustering_distance_rows=sampleDists,
                clustering_distance_cols=sampleDists,
                labels_row = paste(metadata$treatment, 
                                   metadata$genotype, sep = "-"),
                labels_col = metadata$sample_id
                )

ggsave2(filename = "./Plots_OC8901vsOC8345_wt/heat_map_AMANDA.pdf", plot = hm.p)

print(hm.p)

```
**Figure 2:** wt samples cluster by treatment.

### PCA

```{r}
# PCA
principal_components <- c(1,2)
pca.p <- plot_PCA(object = dds.vst, 
                    ntop = 500,
                    color_by = "treat_geno", 
                    shape_by = "experiment",
                    pcs = principal_components,
                    returnData = FALSE)

ggsave(paste0("Plots_OC8901vsOC8345_wt/pca_by_group_PC",principal_components[1],principal_components[2],"_AMANDA.pdf"), plot = pca.p)
print(pca.p)
```
**Figure 3:** Samples group mainly by alpha factor treatment (PC1), but is seems there is a batch effect based on experiment number (PC2).


### Print DESeq2 coefficients

```{r}
print(resultsNames(dds))
```

### Differential expression analysis -wt:

**1- The effect of treatment across all wt samples (the main effect) **

*Note: For additional information about different ways of extracting results from a DESeq object see results() function help by typing ?results in the console,*

```{r, warning=FALSE}
# Create results directory where to store result tables
dir.create("./Results_wt_OC8901vsOC8345", showWarnings = FALSE)

DE_results_list <- list()

coef <- c("treat_geno_alpha_factor_wt_vs_none_wt")

# Set coef_id with the DESeq2 coefficient of interest
coef_id <- "af_vs_none_wt"

# Get DESeq2 results. Note that when using coefficient names, you have to convert the coef vector to a list by doing list(c(coef))
DE_results <- results(dds, contrast=list(c(coef)))

# Shrink Log2FC of genes with low expression 
DE_results <- lfcShrink(dds, 
                       contrast = list(c(coef)),
                       type = "ashr", 
                       res=DE_results )

# Sort table by adjusted p,value (ascending)
DE_results <- DE_results[order(DE_results$padj),]

# Add gene names to results table
DE_results$gene_names <- get_gene_names_from_gene_ids(orf_ids = rownames(DE_results), annotation_db = org.Sc.sgd.db, look_for = "ORF", fetch = "GENENAME")

# Store results table into DE_results list
DE_results_list[[coef_id]] <- DE_results

# Plot top10 diff expressed genes
plot_expression_topN_genes(DE_results_list[[coef_id]], N=5)

summary(dds)
```

### Save tables with results stored in the DE_results_list object

```{r}
for (example_name in names(DE_results_list)){
  # Save results table into a file
  write.table(x = as.data.frame(DE_results_list[[example_name]]), 
            file = paste0("./Results_wt_OC8901vsOC8345/DE_",example_name,".txt"),
            sep = "\t", 
            col.names = NA)
}
```


### Gene enrichment analysis
```{r}
# Aux function to convert string of ORF IDs to string of gene symbol IDs.
get_gene_names <- function(A){
  if(length(A) > 0){
    my_orf_ids <- stringr::str_split(string = A, pattern = "/", simplify = T) %>% 
    as.vector(.)
    get_gene_names_from_gene_ids(orf_ids = my_orf_ids, 
                                 annotation_db = org.Sc.sgd.db, 
                                 look_for = "ORF", fetch = "GENENAME" ) %>%
      stringr::str_flatten(collapse = "/")
  }
  else {
    Return("None")
  }
}

dir.create(path = "./GO_annotation", showWarnings = F)
dir.create(path = "./GO_overrepresentation", showWarnings = F)
dir.create(path = "./GO_gsea", showWarnings = F)

coef_id <- "af_vs_none_wt"
DE_results <- DE_results_list[[coef_id]]

# Sort DE_results table by Log2FC in decreasing order
DE_results.sorted <- DE_results[order(DE_results$log2FoldChange, decreasing = T),]

# Fetch GO gene annotations
go.annot <- groupGO(keyType = "ORF",
               gene     = rownames(DE_results.sorted),
               OrgDb    = org.Sc.sgd.db,
               ont      = "BP",
               level    = 3,
               readable = FALSE) # org.Sc.sgd.db incompatible with this function (SYMBOL col is names GENENAME)

go.annot.df <- as.data.frame(go.annot)


# Save results table into a file
write.table(x = go.annot.df, 
          file = paste0("./GO_annotation/GO_annot_",coef_id,".txt"),
          sep = "\t", 
          col.names = NA)
  

# Overrepresentation analysis of genes with Log2FC >= 1 and padj < 0.05
go_overrep.up <- enrichGO(keyType = "ORF",
                gene          = rownames(subset(DE_results.sorted, padj < 0.05 & log2FoldChange >= 1)),
                #universe      = rownames(DE_results.sorted),
                OrgDb         = org.Sc.sgd.db,
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.05,
                readable      = FALSE)

go_overrep.up.df <- as.data.frame(go_overrep.up)

# Add gene names to results
go_overrep.up.df$gene_name <- unname(unlist(sapply(go_overrep.up.df$geneID, function(x){get_gene_names(x)})))

# Save results table into a file
write.table(x = go_overrep.up.df, 
          file = paste0("./GO_overrepresentation/GO_overrep_",coef_id,"_up.txt"),
          sep = "\t", 
          col.names = NA)

go_overrep.down <- enrichGO(keyType = "ORF",
                gene          = rownames(subset(DE_results.sorted, padj < 0.05 & log2FoldChange <= -1)),
                #universe      = rownames(DE_results.sorted),
                OrgDb         = org.Sc.sgd.db,
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.05,
                readable      = FALSE)

go_overrep.down.df <- as.data.frame(go_overrep.down)

# Add gene names to results
go_overrep.down.df$gene_name <- unname(unlist(sapply(go_overrep.down.df$geneID, function(x){get_gene_names(x)})))

# Save results table into a file
write.table(x = go_overrep.down.df, 
          file = paste0("./GO_overrepresentation/GO_overrep_",coef_id,"_down.txt"),
          sep = "\t", 
          col.names = NA)

# Gene set enrichment analysis (genes have to be provided as a list)
# where names are gene IDs and values are Log2FCs.
geneList <- DE_results.sorted$log2FoldChange
names(geneList) <- rownames(DE_results.sorted)

go.gsea <- gseGO(keyType = "ORF",
              geneList     = geneList,
              OrgDb        = org.Sc.sgd.db,
              ont          = "BP",
              minGSSize    = 10,
              maxGSSize    = 500,
              pvalueCutoff = 0.05,
              verbose      = FALSE)

go.gsea.df <- as.data.frame(go.gsea)

# Add gene names to results
go.gsea.df$gene_name <- unname(unlist(sapply(go.gsea.df$core_enrichment, function(x){get_gene_names(x)})))

# Save results table into a file
write.table(x = go.gsea.df, 
          file = paste0("./GO_gsea/GO_gsea_",coef_id,".txt"),
          sep = "\t", 
          col.names = NA)

```

#### Functions for plotting enrichment results
```{r plot_enrichment_results}
# Enrichment Plotting functions (overrepresentation analysis)
draw_enrichment_barplot <- function(my_enricher_obj, 
                                    my_pathway_counts = 10, 
                                    file_name_prefix = "enricher_barplot", 
                                    my_width = 11, my_height = 8){
  my_enricher_obj@result$p.adjust <- as.numeric(format(my_enricher_obj@result$p.adjust,  digits=3))
  sp <- ggbarplot(my_enricher_obj@result[1:my_pathway_counts, ], 
                  x = "Description", 
                  y = "Count",
                  fill = "p.adjust",          # change fill color by cyl
                  color = "white",            # Set bar border colors to white
                  sort.val = "desc",          # Sort the value in dscending order
                  sort.by.groups = FALSE,     # Don't sort inside each group
                  x.text.angle = 90,          # Rotate vertically x axis texts
                  rotate = TRUE,
                  ggtheme = theme_minimal(),
                  ylab = c("Gene counts"),
  ) + gradient_fill("RdYlBu")
  
  print(sp)
  
  ggsave2(filename = paste0(file_name_prefix,"_barplot_OA.pdf"), plot = sp, width = my_width, height = my_height)
  
  return(sp)
}

draw_enrichment_dotplot<- function(my_enricher_obj, 
                                    my_pathway_counts = 10, 
                                    file_name_prefix = "enricher_dotplot", 
                                    my_width = 11, my_height = 8){
  my_enricher_obj@result$p.adjust <- as.numeric(format(my_enricher_obj@result$p.adjust,  digits=3))
  my_enricher_obj.df <- as.data.frame(my_enricher_obj)
  sp <- dotplot(my_enricher_obj, showCategory= my_pathway_counts) + 
          ggtitle(paste0(file_name_prefix," dotplot")) + 
          theme(axis.text=element_text(size=3))
  
  print(sp)
  
  ggsave2(filename = paste0(file_name_prefix,"_dotplot_OA.pdf"), 
          plot = sp, width = my_width, height = my_height)
  
  return(sp)
}


draw_enrichment_network<- function(dds_res, my_enricher_obj, 
                                    my_pathway_counts = 400, 
                                    file_name_prefix = "enricher_dotplot", 
                                    my_width = 11, my_height = 8){
  
  # Save Log2FC vals in log2fc_symb.df
  log2fc_symb.df <- select(as.data.frame(dds_res), c("log2FoldChange","padj"))
  
  log2fc_symb.df$symbols <- get_gene_names_from_gene_ids(orf_ids = rownames(log2fc_symb.df),
                                                         annotation_db = org.Sc.sgd.db, 
                                                         look_for = "ORF", 
                                                         fetch = "GENENAME", 
                                                         return_all = TRUE)
  
  distance_matrix <- pairwise_termsim(my_enricher_obj, showCategory = my_pathway_counts)
            
  # Get median Log2FC per identified pathway
  my_median_log2fc <- c()
  
  for (g in distance_matrix@result$geneID){
    g.vec <- strsplit(g, "/")[[1]]
    log2fc.median <- median(subset(log2fc_symb.df, rownames(log2fc_symb.df) %in% g.vec)[,"log2FoldChange"])
    my_median_log2fc <- c(my_median_log2fc,log2fc.median)
  }
  
  # Add median Log2FC column
  if(length(my_median_log2fc) == 0){my_median_log2fc <- 0}
  distance_matrix@result$median.log2fc <- my_median_log2fc
      
  # Network plot
  p6 <- emapplot(distance_matrix, 
                 repel = T, 
                 showCategory = my_pathway_counts, 
                 legend_n = 5, 
                 min_edge = 0.4 , 
                 color = "median.log2fc", 
                 cex_label_category = 0.4,
                 node_label = "category", 
                 label_format = 20)
  
  
  
  print(p6)
  
  ggsave2(filename = paste0(file_name_prefix,"_network_OA.pdf"), 
          plot = p6, width = my_width, height = my_height)
  
  return(p6)
}

draw_enrichment_tree<- function(dds_res, my_enricher_obj, 
                                    my_pathway_counts = 400, 
                                    file_name_prefix = "enricher_dotplot", 
                                    my_width = 11, my_height = 8){
  
  # Save Log2FC vals in log2fc_symb.df
  log2fc_symb.df <- select(as.data.frame(dds_res), c("log2FoldChange","padj"))
  
  log2fc_symb.df$symbols <- get_gene_names_from_gene_ids(orf_ids = rownames(log2fc_symb.df),
                                                         annotation_db = org.Sc.sgd.db, 
                                                         look_for = "ORF", 
                                                         fetch = "GENENAME", 
                                                         return_all = TRUE)
  
  distance_matrix <- pairwise_termsim(my_enricher_obj, showCategory = my_pathway_counts)
            
  # Get median Log2FC per identified pathway
  my_median_log2fc <- c()
  
  for (g in distance_matrix@result$geneID){
    g.vec <- strsplit(g, "/")[[1]]
    log2fc.median <- median(subset(log2fc_symb.df, rownames(log2fc_symb.df) %in% g.vec)[,"log2FoldChange"])
    my_median_log2fc <- c(my_median_log2fc,log2fc.median)
  }
  
  # Add median Log2FC column
  if(length(my_median_log2fc) == 0){my_median_log2fc <- 0}
  distance_matrix@result$median.log2fc <- my_median_log2fc
      
  # Treeplots
  number_of_categories = min(80, as.vector(table(distance_matrix@result$p.adjust < 0.05)[['TRUE']]))
  p7 <- treeplot(distance_matrix, 
                showCategory = 80, 
                nCluster = round(2 * sqrt(number_of_categories), 
                                 digits = 0), 
                color = "median.log2fc", 
                nWords = 0)
  
  
  
  print(p7)
  
  ggsave2(filename = paste0(file_name_prefix,"_tree_OA.pdf"), 
          plot = p7, width = my_width, height = my_height)
  
  return(p7)
}

# Enrichment plotting functions (gsea analysis)
draw_GSEA_barplot <- function(my_gsea_obj, 
                              my_pathway_counts = 10, 
                              file_name_prefix = "gsea_barplot", 
                              my_width = 11, my_height = 8){
  my_gsea_obj@result$p.adjust <- as.numeric(format(my_gsea_obj@result$p.adjust,  digits=3))
  sp <- ggbarplot(head(my_gsea_obj@result, n = my_pathway_counts), 
                  x = "Description", 
                  y = "NES",
                  fill = "p.adjust",          # change fill color by cyl
                  color = "white",            # Set bar border colors to white
                  sort.val = "desc",          # Sort the value in dscending order
                  sort.by.groups = FALSE,     # Don't sort inside each group
                  x.text.angle = 90,          # Rotate vertically x axis texts
                  rotate = TRUE,
                  ggtheme = theme_minimal(),
                  ylab = c("Normalized Enrichment Score (NES)"),
                  lab.size = 3
  ) + gradient_fill("RdYlBu")
  print(sp)
  ggsave2(filename = paste0(file_name_prefix,"_barplot_GSEA.pdf"), plot = sp, width = my_width, height = my_height)
  return(sp)
}

draw_GSEA_dotplot<- function(my_enricher_obj, # test
                                    my_pathway_counts = 10, 
                                    file_name_prefix = "enricher_dotplot", 
                                    my_width = 11, my_height = 8){
  my_enricher_obj@result$p.adjust <- as.numeric(format(my_enricher_obj@result$p.adjust,  digits=3))
  my_enricher_obj.df <- as.data.frame(my_enricher_obj)
  sp <- dotplot(my_enricher_obj, x = "NES",
                showCategory= my_pathway_counts) + 
          ggtitle(paste0(file_name_prefix," dotplot")) + 
          theme(axis.text=element_text(size=3))
  
  print(sp)
  
  ggsave2(filename = paste0(file_name_prefix,"_dotplot_GSEA.pdf"), 
          plot = sp, width = my_width, height = my_height)
  
  return(sp)
}


draw_GSEA_network<- function(my_enricher_obj, # test 
                                    my_pathway_counts = 400, 
                                    file_name_prefix = "enricher_dotplot", 
                                    my_width = 11, my_height = 8){
  
  
  distance_matrix <- pairwise_termsim(my_enricher_obj, showCategory = my_pathway_counts)
  
  # Network plot
  p6 <- emapplot(distance_matrix, 
                 repel = T, 
                 showCategory = my_pathway_counts, 
                 legend_n = 5, 
                 min_edge = 0.4 , 
                 color = "NES", 
                 cex_label_category = 0.4,
                 node_label = "category", 
                 label_format = 20)
  
  
  
  print(p6)
  
  ggsave2(filename = paste0(file_name_prefix,"_network_GSEA.pdf"), 
          plot = p6, width = my_width, height = my_height)
  
  return(p6)
}

draw_GSEA_tree<- function(my_enricher_obj, # test 
                                    my_pathway_counts = 400, 
                                    file_name_prefix = "enricher_dotplot", 
                                    my_width = 11, my_height = 8){
  
  distance_matrix <- pairwise_termsim(my_enricher_obj, showCategory = my_pathway_counts)
      
  # Treeplots
  number_of_categories = min(80, 
                             nrow(subset(distance_matrix@result, p.adjust < 0.05)),
                             my_pathway_counts)
                            
  p7 <- treeplot(distance_matrix, 
                showCategory = number_of_categories, 
                nCluster = round(2 * sqrt(number_of_categories), digits = 0), 
                color = "NES", 
                nWords = 0,
                hilight = FALSE,
                legend_n = 7)
  
  print(p7)
  
  ggsave2(filename = paste0(file_name_prefix,"_tree_GSEA.pdf"), 
          plot = p7, width = my_width, height = my_height)
  
  return(p7)
}





```


### Plotting enrichment results using the functions above.
```{r}
# Overrepresentation analysis upregulated genes
p1 <- draw_enrichment_barplot(my_enricher_obj = go_overrep.up, my_pathway_counts = 1000, file_name_prefix = "./GO_overrepresentation/GO_af_vs_none_wt_UP", my_width = 10, my_height = 10)

# Dotplot
p2 <- draw_enrichment_dotplot(my_enricher_obj = go_overrep.up,  file_name_prefix = "./GO_overrepresentation/GO_af_vs_none_wt_UP", my_width = 10, my_height = 10)

p3 <- draw_enrichment_network(dds_res = DE_results_list[["af_vs_none_wt"]], my_enricher_obj = go_overrep.up, my_pathway_counts = 200, file_name_prefix = "./GO_overrepresentation/GO_af_vs_none_wt_UP", my_width = 10, my_height = 20)

p4 <- draw_enrichment_tree(dds_res = DE_results_list[["af_vs_none_wt"]], my_enricher_obj = go_overrep.up, my_pathway_counts = 200, file_name_prefix = "./GO_overrepresentation/GO_af_vs_none_wt_UP", my_width = 10, my_height = 20)

# Overrepresentation analysis downregulated genes

p5 <- draw_enrichment_barplot(my_enricher_obj = go_overrep.down, my_pathway_counts = 30, file_name_prefix = "./GO_overrepresentation/GO_af_vs_none_wt_DOWN", my_width = 10, my_height = 10)

# Dotplot
p6 <- draw_enrichment_dotplot(my_enricher_obj = go_overrep.down, my_pathway_counts = 10, file_name_prefix = "./GO_overrepresentation/GO_af_vs_none_wt_DOWN", my_width = 10, my_height = 10)

p7 <- draw_enrichment_network(dds_res = DE_results_list[["af_vs_none_wt"]], my_enricher_obj = go_overrep.down, my_pathway_counts = 200, file_name_prefix = "./GO_overrepresentation/GO_af_vs_none_wt_DOWN", my_width = 10, my_height = 20)

p8 <- draw_enrichment_tree(dds_res = DE_results_list[["af_vs_none_wt"]], my_enricher_obj = go_overrep.down, my_pathway_counts = 200, file_name_prefix = "./GO_overrepresentation/GO_af_vs_none_wt_DOWN", my_width = 10, my_height = 20)


# GSEA

p9 <- draw_GSEA_barplot(my_gsea_obj = go.gsea, file_name_prefix = "./GO_gsea/GO_af_vs_none_wt", my_width = 10, my_height = 10, my_pathway_counts = 30)

p10 <- draw_GSEA_dotplot(my_enricher_obj = go.gsea, my_pathway_counts = 10, file_name_prefix = "./GO_gsea/GO_af_vs_none_wt", my_width = 10, my_height = 10) ## CHECK color assignment

p11 <- draw_GSEA_network(my_enricher_obj = go.gsea, my_pathway_counts = 200, file_name_prefix = "./GO_gsea/GO_af_vs_none_wt", my_width = 10, my_height = 20) ## CHECK Log2FC info

p12 <- draw_GSEA_tree(my_enricher_obj = go.gsea, my_pathway_counts = 200, file_name_prefix = "./GO_gsea/GO_af_vs_none_wt", my_width = 10, my_height = 20) ## CHECK Log2FC info


```



