---
title: "Over-representation analysis GO BP - ime4D after growth genes removed"
author: "Amanda Yeo, Jason Rogers, Hernan Lorenzi, Orna Cohen-Fix"
date: "2024-06-17"
output: html_document
---

````{r include=FALSE}
knitr::opts_chunk$set(message = FALSE, 
                      warning = FALSE,
                      tidy.opts=list(width.cutoff=80),
                      tidy=TRUE
                      )
```

### Install required packages

```{r}
### Install required R libraries/packages:

# CRAN repository:
required_cran_packages <- c("cowplot","ggpubr","ggsci","RColorBrewer","pheatmap")
new_cran_packages <- required_cran_packages[!(required_cran_packages %in%
                                                installed.packages()[,"Package"])]
if(length(new_cran_packages)>0) {install.packages(new_cran_packages)}

# Bioconductor repository:
required_bioconductor_packages <- c("DESeq2","org.Sc.sgd.db","EnhancedVolcano","enrichplot","clusterProfiler")

new_bioconductor_packages <- required_bioconductor_packages[!(required_bioconductor_packages %in%
                                                installed.packages()[,"Package"])]

if(length(new_bioconductor_packages)>0) {
  if (!require("BiocManager", quietly = TRUE)){
    install.packages("BiocManager")
  }
  BiocManager::install(new_bioconductor_packages)
}
```

### Load R libraries

```{r}
# Packages for differential expression analysis and plotting:

suppressWarnings(
  suppressPackageStartupMessages({
    library(DESeq2);  # Package for differential expression analysis
    library(cowplot); # Package for saving plots (ggsave2)
    library(ggpubr); # Package for plotting functions
    library(ggsci);
    library(RColorBrewer); 
    library(pheatmap); # Package for volcane plot functions
    library(org.Sc.sgd.db); # Yeast annotations
    library(EnhancedVolcano);
    library(clusterProfiler);
    library(enrichplot);
    }
  )
)
# Packages for functional analysis:

#library(clusterProfiler)
#library(enrichplot)
#library(ggupset)
#library(msigdbr)
#library(ReactomePA)
#library(ggraph)
#library(BSgenome)
```

### Load useful functions

```{r}
#
# --- function for adding gene names to DESeq2 result table ---
#
get_gene_names_from_gene_ids <- function(orf_ids, annotation_db, 
                                         look_for = 'ORF', 
                                         fetch = 'GENENAME', return_all = TRUE){
  # Reference organism: Saccharomyces cerevisiae => DATABASE = org.Sc.sgd.db
  symbols <- mapIds(annotation_db, keys = orf_ids, column = fetch, 
                    keytype = look_for, multiVals = "first")
  symbols <- symbols[!is.na(symbols)]
  to_name <- orf_ids %in% names(symbols)
  orf_ids[to_name] <- as.vector(symbols)
  if (return_all){
    return(orf_ids)
  }
  else {
    return(orf_ids[to_name])
  }
}

#
# --- function for PCA plots ---
#
plot_PCA = function(object, color_by="condition", 
                    shape_by = 19, ntop=500, size = 3,
                    returnData=FALSE, pcs = c(1,2))
{
  # Check variables are present in object
  intgroup = c(color_by)
  if (shape_by != 19){intgroup <- c(intgroup, shape_by)}
  if (!all(intgroup %in% names(colData(object)))) {
    stop("the argument 'intgroup' should specify columns of colData(dds)")
  }

  # calculate the variance for each gene
  rv <- rowVars(assay(object))

  # select the ntop genes by variance
  select <- order(rv, decreasing=TRUE)[seq_len(min(ntop, length(rv)))]

  # perform a PCA on the data in assay(x) for the selected genes
  pca <- prcomp(t(assay(object)[select,]))

  # the contribution to the total variance for each component
  percentVar <- pca$sdev^2 / sum( pca$sdev^2 )

  
  intgroup.df <- as.data.frame(colData(object)[, intgroup, drop=FALSE])
  
  # add the intgroup factors together to create a new grouping factor
  group <- if (length(intgroup) > 1) {
    factor(apply( intgroup.df, 1, paste, collapse=":"))
  } else {
    colData(object)[[intgroup]]
  }

  # assembly the data for the plot
  d <- data.frame(PC1=pca$x[,pcs[1]], PC2=pca$x[,pcs[2]], group=group, intgroup.df, name=colnames(object))
  colnames(d)[1] <- paste0("PC",pcs[1])
  colnames(d)[2] <- paste0("PC",pcs[2])
  
  if (returnData) {
    attr(d, "percentVar") <- percentVar[1:2]
    return(d)
  }
  
  ggplot(data=d, aes_string(x=colnames(d)[1], y=colnames(d)[2], color=color_by, shape=shape_by)) + 
    geom_point(size=size) + 
    scale_color_lancet() +
    xlab(paste0("PC",pcs[1],": ",round(percentVar[pcs[1]] * 100),"% variance")) + # fixed
    ylab(paste0("PC",pcs[2],": ",round(percentVar[pcs[2]] * 100),"% variance")) + # fixed
    coord_fixed(ratio = (max(d[,1])-min(d[,1]))/(max(d[,2])-min(d[,2])))
}

#
# --- function to plot gene expression of top10 genes ---
#
# Plot top10 diff expressed genes
plot_expression_topN_genes <- function(deseq_result, N=10){
  my_top10 <- head(rownames(deseq_result),N)
  for (gene in my_top10){
    print(gene)
    plotCounts(dds, gene, 
               intgroup = c("treatment","genotype"), 
               returnData = FALSE, col=metadata$genotype)
  }
}
```


```{r}
ime4_up <- read.table(file = "DeUp_ime4D_growthgenesrem.txt", header = TRUE, row.names = 1, sep = "\t")
ime4_down <- read.table(file = "DEdown_ime4D_growthgenesrem.txt", header = TRUE, row.names = 1, sep = "\t")

ime4_rownames_up <- rownames(ime4_up)
ime4_rownames_down <- rownames(ime4_down)
```

```{r}
# Aux function to convert string of ORF IDs to string of gene symbol IDs.
get_gene_names <- function(A){
  if(length(A) > 0){
    my_orf_ids <- stringr::str_split(string = A, pattern = "/", simplify = T) %>% 
    as.vector(.)
    get_gene_names_from_gene_ids(orf_ids = my_orf_ids, 
                                 annotation_db = org.Sc.sgd.db, 
                                 look_for = "ORF", fetch = "GENENAME" ) %>%
      stringr::str_flatten(collapse = "/")
  }
  else {
    Return("None")
  }
}

dir.create(path = "./GO_annotation", showWarnings = F)
dir.create(path = "./GO_overrepresentation", showWarnings = F)
# Sort DE_results table by Log2FC in decreasing order
DE_resultsUP.sorted <- ime4_up

# Fetch GO gene annotations
go.annot <- groupGO(keyType = "ORF",
               gene     = rownames(DE_resultsUP.sorted),
               OrgDb    = org.Sc.sgd.db,
               ont      = "BP",
               level    = 3,
               readable = FALSE) # org.Sc.sgd.db incompatible with this function (SYMBOL col is names GENENAME)

go.annot.df <- as.data.frame(go.annot)


# Save results table into a file
write.table(x = go.annot.df, 
          file = paste0("./GO_annotation/GO_annot_DEUP_ime4D_growthgenesrem.txt"),
          sep = "\t", 
          col.names = NA)
  

# Overrepresentation analysis of genes with Log2FC >= 1 and padj < 0.05
go_overrep.up <- enrichGO(keyType = "ORF",
                gene = rownames(subset(DE_resultsUP.sorted)),
                #universe      = rownames(DE_results.sorted),
                OrgDb         = org.Sc.sgd.db,
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.05,
                readable      = FALSE)

go_overrep.up.df <- as.data.frame(go_overrep.up)

# Add gene names to results
go_overrep.up.df$gene_name <- unname(unlist(sapply(go_overrep.up.df$geneID, function(x){get_gene_names(x)})))

# Save results table into a file
write.table(x = go_overrep.up.df, 
          file = paste0("./GO_overrepresentation/GO_overrep_OC9501ime4_DEup_growthgenesrem.txt"),
          sep = "\t", 
          col.names = NA)
```

```{r}
# Aux function to convert string of ORF IDs to string of gene symbol IDs.
get_gene_names <- function(A){
  if(length(A) > 0){
    my_orf_ids <- stringr::str_split(string = A, pattern = "/", simplify = T) %>% 
    as.vector(.)
    get_gene_names_from_gene_ids(orf_ids = my_orf_ids, 
                                 annotation_db = org.Sc.sgd.db, 
                                 look_for = "ORF", fetch = "GENENAME" ) %>%
      stringr::str_flatten(collapse = "/")
  }
  else {
    Return("None")
  }
}

dir.create(path = "./GO_annotation", showWarnings = F)
dir.create(path = "./GO_overrepresentation", showWarnings = F)
# Sort DE_results table by Log2FC in decreasing order
DE_resultsDOWN.sorted <- ime4_down

# Fetch GO gene annotations
go.annotdown <- groupGO(keyType = "ORF",
               gene     = rownames(DE_resultsDOWN.sorted),
               OrgDb    = org.Sc.sgd.db,
               ont      = "BP",
               level    = 3,
               readable = FALSE) # org.Sc.sgd.db incompatible with this function (SYMBOL col is names GENENAME)

go.annot.down.df <- as.data.frame(go.annotdown)


# Save results table into a file
write.table(x = go.annot.down.df, 
          file = paste0("./GO_annotation/GO_annot_DOWN_ime4D_growthgenesrem.txt"),
          sep = "\t", 
          col.names = NA)
  

# Overrepresentation analysis of genes with Log2FC >= 1 and padj < 0.05
go_overrep.down <- enrichGO(keyType = "ORF",
                gene = rownames(subset(DE_resultsDOWN.sorted)),
                #universe      = rownames(DE_results.sorted),
                OrgDb         = org.Sc.sgd.db,
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.05,
                readable      = FALSE)

go_overrep.down.df <- as.data.frame(go_overrep.down)

# Add gene names to results
go_overrep.down.df$gene_name <- unname(unlist(sapply(go_overrep.down.df$geneID, function(x){get_gene_names(x)})))

# Save results table into a file
write.table(x = go_overrep.down.df, 
          file = paste0("./GO_overrepresentation/GO_overrep_OC9501ime4_DEdown_growthgenesrem.txt"),
          sep = "\t", 
          col.names = NA)
```


#### Functions for plotting enrichment results
```{r plot_enrichment_results}
# Enrichment Plotting functions (overrepresentation analysis)

##### Add p-value cutoff adjp cutoff
draw_enrichment_barplot <- function(my_enricher_obj, 
                                    my_pathway_counts = 10, 
                                    file_name_prefix = "enricher_barplot", 
                                    my_width = 11, 
                                    my_height = 8, 
                                    padj_cutoff = 1){
  
  my_results <- my_enricher_obj@result
  
  # Subset 
  my_results <- subset(my_results, p.adjust <= padj_cutoff)
  
  # Format p.adj values with 3 digits only
  my_results$p.adjust <- as.numeric(format(my_results$p.adjust,  digits=3))
  
  # Keep top my_pathway_counts gene sets
  my_results <- head(my_results, my_pathway_counts)
  
  sp <- ggbarplot(my_results, 
                  x = "Description", 
                  y = "Count",
                  fill = "p.adjust",          # change fill color by cyl
                  color = "white",            # Set bar border colors to white
                  sort.val = "desc",          # Sort the value in dscending order
                  sort.by.groups = FALSE,     # Don't sort inside each group
                  x.text.angle = 90,          # Rotate vertically x axis texts
                  rotate = TRUE,
                  ggtheme = theme_minimal(),
                  ylab = c("Gene counts"),
  ) + gradient_fill("RdYlBu")
  
  print(sp)
  
  ggsave2(filename = paste0(file_name_prefix,"_barplot_OA.pdf"), plot = sp, width = my_width, height = my_height)
  
  return(sp)
}
```


### Plotting enrichment results using the functions above.
```{r}

# Overrepresentation analysis upregulated genes
p1 <- draw_enrichment_barplot(my_enricher_obj = go_overrep.up, my_pathway_counts = 300, file_name_prefix = "./GO_overrepresentation/GO_OC9501ime42_UP_growthgenesrem", my_width = 10, my_height = 10, padj_cutoff = 0.05)

p2 <- draw_enrichment_barplot(my_enricher_obj = go_overrep.down, my_pathway_counts = 300, file_name_prefix = "./GO_overrepresentation/GO_OC9501ime4_DOWN_growthgenesrem", my_width = 10, my_height = 10, padj_cutoff = 0.05)
```